sematic/abstract_future.py

```diff
@@ -81,9 +83,51 @@ class FutureProperties:
     inline: bool
     name: str
     tags: List[str]
+    kwargs: Dict[str, Any]
     resource_requirements: Optional[ResourceRequirements] = None
     retry_settings: Optional[RetrySettings] = None
     base_image_tag: Optional[str] = None
+    external_resources: List[ExternalResource] = field(default_factory=list)
+
+    def get_resolved_kwargs(self) -> Dict[str, Any]:
+        """
+        Extract only resolved/concrete kwargs.
+
+        For external resources, only ones in the ACTIVE state can be used
+        and are considered resolved.
+        """
+        resources_by_id = {r.id: r for r in self.external_resources}
+        resolved_kwargs = {}
+        for name, value in self.kwargs.items():
+            if isinstance(value, AbstractFuture):
+                if value.state == FutureState.RESOLVED:
+                    resolved_kwargs[name] = value.value
+            elif isinstance(value, ExternalResource):
+                value = resources_by_id.get(value.id)
+                if value.status.state == ResourceState.ACTIVE:
+                    resolved_kwargs[name] = value
+            else:
+                resolved_kwargs[name] = value
+
+        return resolved_kwargs
+
+    def all_args_ready(self) -> bool:
+        """True if all args are sufficiently ready for the future to be scheduled.
+
+        External resources are scheduled along with the future and thus need not be
+        active for this to return True.
+        """
+        missing_kwarg_keys = set(self.kwargs.keys()).difference(
+            self.get_resolved_kwargs().keys()
+        )
+        for key in missing_kwarg_keys:
+            value = self.kwargs[key]
+            if not isinstance(value, ExternalResource):
+                # it's ok if there are args that aren't usable yet
+                # if they're external resources. Those will get scheduled
+                # along with the future.
```

Comment:
Why not remove the `if value.status.state == ResourceState.ACTIVE` from `get_resolved_kwargs` since their status does not matter?